File: Python/Dynamic_Programming/fibonacci_memoization.py
#   FibonacciMemoization : Overlapping Subproblems Property in Dynamic_Programming

# to memoize the value of the pth fibbonacci number, where p < n
fibb_memo = dict()


def fibonacciMemoization(n: int):
    """
    :param n: range numbers to be visited in fibonacci series
    :return: nth value in the fibonacci sequence
    """

    global fibb_memo

    if n in fibb_memo:
        return fibb_memo[n]

    if n <= 2:
        return 1

    fibb_memo[n] = fibonacciMemoization(n - 1) + fibonacciMemoization(n - 2)

    return fibb_memo[n]


if __name__ == '__main__':
    num = int(input("Enter a number: "))
    print("Fib: ", fibonacciMemoization(num))

File: Golang/Dynamic Programming/fibonacci_memoization.go
package main

import "fmt"

func main() {
	var num int
	fmt.Println("Enter a number: ")
	_, _ = fmt.Scanf("%d", &num)
	// fibValues := make([]int, num)
	fib := fibb(num)
	fmt.Println(fib)
}

func fibb(num int) int {
	fib := make([]int, num)
	if fib[num] != 0 {
		return fib[num]
	}

	if num <= 2 {
		return 1
	}

	fib[num] = fibb(num - 1) + fibb(num - 2)

	return fib[num]
}
